tools/migration/internal/runner.go

```diff
@@ -27,22 +40,112 @@ type Migration interface {
+	// verbose controls the amount of output.
+	verbose bool
+
+	// command is the migration command to run, passed from the CLI
+	command string
+
+	// oldRepoOpt is the option passed  from the CLI.
 	oldRepoOpt string
+
+	// MigrationsProvider is a dependency for fetching available migrations
+	// to allow unit tests to supply test migrations without creating test fixtures.
+	MigrationsProvider func() []Migration
+
+	// RepoVersionGetter is a dependency for getting a Repo Version,
+	// to allow the binary version of the repo to be mocked in testing.
+	RepoVersionGetter func() uint
 }
 
 func NewMigrationRunner(verb bool, command, oldRepoOpt string) *MigrationRunner {
-	// TODO: Issue #2585 Implement repo migration version detection and upgrade decisioning
-
 	return &MigrationRunner{
-		verbose:    verb,
-		command:    command,
-		oldRepoOpt: oldRepoOpt,
+		verbose:            verb,
+		command:            command,
+		oldRepoOpt:         oldRepoOpt,
+		MigrationsProvider: DefaultMigrationsProvider,
+		RepoVersionGetter:  DefaultVersionGetter,
 	}
 }
 
+// DefaultVersionGetter fetches the Version constant from the repo package.
+func DefaultVersionGetter() uint {
+	return repo.Version
+}
+
 func (m *MigrationRunner) Run() error {
 	// TODO: Issue #2595 Implement first repo migration
+	repoVersion, err := m.loadVersion()
+	if err != nil {
+		return err
+	}
+	if repoVersion == repo.Version {
+		return errors.New("binary version = repo version; migration not run")
+	}
+	for _, mig := range m.MigrationsProvider() {
+		from, to := mig.Versions()
+		if from == repoVersion && to == from+1 {
+			newRepoPath, err := CloneRepo(m.oldRepoOpt)
+			if err != nil {
+				return err
+			}
+
+			switch m.command {
+			case "migrate":
+				if err = mig.Migrate(newRepoPath); err != nil {
+					return err
+				}
+				if err = m.validateAndUpdateVersion(to, newRepoPath, mig); err != nil {
+					return err
+				}
+				if err = InstallNewRepo(m.oldRepoOpt, newRepoPath); err != nil {
+					return err
+				}
+			case "buildonly":
+				if err = mig.Migrate(newRepoPath); err != nil {
+					return err
+				}
+			case "install":
+				if err = m.validateAndUpdateVersion(to, newRepoPath, mig); err != nil {
+					return err
+				}
+				if err = InstallNewRepo(m.oldRepoOpt, newRepoPath); err != nil {
+					return err
+				}
+			}
+			return nil
+		}
+	}
+	return fmt.Errorf("did not find valid repo migration for version %d to version %d", repoVersion, repoVersion+1)
+}
+
+// Shamelessly lifted from FSRepo, with version checking added.
+func (m *MigrationRunner) loadVersion() (uint, error) {
+	file, err := ioutil.ReadFile(filepath.Join(m.oldRepoOpt, "version"))
+	if err != nil {
+		return 0, err
+	}
+
+	strVersion := strings.Trim(string(file), "\n")
+	version, err := strconv.Atoi(strVersion)
+	if err != nil {
+		return 0, err
+	}
+
+	if version < 0 || version > 10000 {
+		return 0, fmt.Errorf("repo version out of range: %s", strVersion)
+	}
+
+	return uint(version), nil
+}
```

Comment:
I think we're better off sharing this code (perhaps as a free function in fsrepo), if possible. Definitely remove duplication of the "version" filename

Reply:
IMO duplication of the "version" filename is WAI.  I think we can assert that this is a well known and agreed upon feature of any filesystem repo just as the [fs-repo spec for IPFS does](https://github.com/ipfs/specs/tree/master/repo/fs-repo): "Any fs-repo implementation of any versions >0 MUST be able to read the version file. It MUST NOT change format between versions."

Reply:
Though I guess keeping a free function in fs-repo for this fits into this world view just as well

Reply:
Added the free function repo.VersionFilename

Reply:
My mistake, I misunderstood what you asked for.  I also still thought we were trying to minimize intrusions into go-filecoin.